### PR TITLE
[10.0][ADD] Sendgrid modules

### PR DESCRIPTION
**Pull request by ecino**
These new modules allow to use Sendgrid as mail provider and benefit from tracking functionalities as well as templating functionalities from Sendgrid.

### CODE REVIEW

File: mail_sendgrid/README.rst
Comment (pedrobaeza) [inline mail_sendgrid/README.rst]: s/9.0/10.0

File: mail_sendgrid/__init__.py
Comment (pedrobaeza) [inline mail_sendgrid/__init__.py]: Please use current template and put correctly `coding`: https://github.com/OCA/maintainer-tools/blob/master/template/module/__openerp__.py#L1-L3

Comment (ecino) [reply mail_sendgrid/__init__.py]: Fixed all headers

**Comment by ecino** [comment]
Don't know why pylint fails on importing lxml... it is installed in .travis.yml file https://github.com/OCA/social/blob/10.0/.travis.yml#L13

**Comment by pedrobaeza** [comment]
It's a general problem: https://github.com/OCA/maintainer-quality-tools/issues/467

**Comment by ecino** [comment]
Tests are fixed, the module is stable, we are using it in production in version 10 since some time now. We only miss a review to merge.

**Comment by ecino** [comment]
Is someone willing to review?

**Comment by pedrobaeza** [comment]
Are you able to improve the coverage for having everything green? If so, I will optimistically merge the PR although no reviews.

**Comment by ecino** [comment]
I already tried to improve the coverage but the targeted value is very high: here is the code putting down the coverage, but I don't know if we could include those parts in the tests and if it would be relevant, what do you think?

https://codecov.io/gh/OCA/social/src/d97bd6f63d54ee8fa029ea8c08ed2482b40571d6/mail_sendgrid/controllers/sendgrid_event_webhook.py

https://codecov.io/gh/OCA/social/src/d97bd6f63d54ee8fa029ea8c08ed2482b40571d6/mail_sendgrid/controllers/json_request.py

**Comment by hbrunn** [comment]
@pedrobaeza why would we do such a thing? Of course no merging without review, and also no merging with red CI, even if it's "only" tests.

**Comment by hbrunn** [comment]
@ecino I'd propose you review some PRs and ask the author of those to return the favor

**Comment by hbrunn** [comment]
...for testing your controllers, you can use [HttpCase](https://github.com/OCA/OCB/blob/11.0/odoo/tests/common.py#L247)

**Comment by pedrobaeza** [comment]
@hbrunn I reviewed his work on previous version and I'll review also before merging this one, but no one is able to review such a big thing, and it's a good work from not a newcomer (@ecino has made already several contributions). It's for not discouraging him from more contributions.

**Comment by hbrunn** [comment]
I understand the reasoning, still I think if we lower our standard everyone loses. Anyways, now realizing most of the code is already reviewed and merged, I could review the rest easily and we've one +1. @ecino what I often do in such cases it to tell the review which commits are the interesting ones, that allows reviewers to trust the rest and focus on new stuff

**Review by hbrunn — APPROVED** [review]

**Comment by ecino** [comment]
@hbrunn Thanks for the tips, it's always useful to improve myself. In that case the relevant commits are the 3 last (after migration of code for version 10)

File: mail_sendgrid/__manifest__.py
Comment (simahawk) [inline mail_sendgrid/__manifest__.py]: https://github.com/OCA/social

File: mail_sendgrid/controllers/sendgrid_event_webhook.py
Comment (simahawk) [inline mail_sendgrid/controllers/sendgrid_event_webhook.py]: one line docstring

File: mail_sendgrid/models/email_lang_template.py
Comment (simahawk) [inline mail_sendgrid/models/email_lang_template.py]: `selection='_select_lang'`

File: mail_sendgrid/models/email_template.py
Comment (simahawk) [inline mail_sendgrid/models/email_template.py]: is this needed?

File: mail_sendgrid/models/email_template.py
Comment (simahawk) [inline mail_sendgrid/models/email_template.py]: `= []`

File: mail_sendgrid/models/email_template.py
Comment (simahawk) [inline mail_sendgrid/models/email_template.py]:
```python
substitution_vals = {}.fromkeys(res_ids, [])
```
or even better
```python
from collection import defaultdict
substitution_vals = defaultdict(list)
```

File: mail_sendgrid/models/email_tracking.py
Comment (simahawk) [inline mail_sendgrid/models/email_tracking.py]: readonly?

File: mail_sendgrid/models/email_tracking.py
Comment (simahawk) [inline mail_sendgrid/models/email_tracking.py]: do you expect many records? if yes, go for a `search` instead of `filtered`

File: mail_sendgrid/models/email_tracking.py
Comment (simahawk) [inline mail_sendgrid/models/email_tracking.py]: better to prefix this w/ `_sendgrid` as you did - nicely - w/ the others (unless this comes from the mail_tracking module ;P)

File: mail_sendgrid/models/email_tracking.py
Comment (simahawk) [inline mail_sendgrid/models/email_tracking.py]: ```except ValueError```

File: mail_sendgrid/models/email_tracking.py
Comment (simahawk) [inline mail_sendgrid/models/email_tracking.py]: default is `None` in any case

File: mail_sendgrid/models/email_tracking.py
Comment (simahawk) [inline mail_sendgrid/models/email_tracking.py]: do you need this further "if"?

File: mail_sendgrid/models/mail_mail.py
Comment (simahawk) [inline mail_sendgrid/models/mail_mail.py]: same as above

File: mail_sendgrid/models/mail_mail.py
Comment (simahawk) [inline mail_sendgrid/models/mail_mail.py]: same story: if you expect many records use a search

File: mail_sendgrid/models/mail_mail.py
Comment (simahawk) [inline mail_sendgrid/models/mail_mail.py]: MailMail

File: mail_sendgrid/models/mail_mail.py
Comment (simahawk) [inline mail_sendgrid/models/mail_mail.py]: use `write` to write once on both fields

File: mail_sendgrid/models/mail_mail.py
Comment (simahawk) [inline mail_sendgrid/models/mail_mail.py]: if possible break this into smallest methods

File: mail_sendgrid/models/mail_tracking_event.py
Comment (simahawk) [inline mail_sendgrid/models/mail_tracking_event.py]: use `Copyright` instead of symbol

File: mail_sendgrid/models/sendgrid_template.py
Comment (simahawk) [inline mail_sendgrid/models/sendgrid_template.py]: :stop_sign: `update` is an[ internal method](https://github.com/odoo/odoo/blob/10.0/odoo/models.py#L5010) and is used to update in memory record w/ writing/creating things.

File: mail_sendgrid/tests/test_mail_sendgrid.py
Comment (simahawk) [inline mail_sendgrid/tests/test_mail_sendgrid.py]: if possible use `setUpClass` to create records that are always used: boosts test perf a lot ;)

File: mail_sendgrid/views/email_template_view.xml
Comment (simahawk) [inline mail_sendgrid/views/email_template_view.xml]: `/ancestor::page` should be better

File: mail_sendgrid/views/sendgrid_email_view.xml
Comment (simahawk) [inline mail_sendgrid/views/sendgrid_email_view.xml]: same here

File: mail_sendgrid/wizards/email_template_preview.py
Comment (simahawk) [inline mail_sendgrid/wizards/email_template_preview.py:14]: `onchange_[...]`

File: mail_sendgrid_mass_mailing/README.rst
Comment (simahawk) [inline mail_sendgrid_mass_mailing/README.rst]: update this w/ [latest README](https://github.com/OCA/maintainer-tools/blob/master/template/module/README.rst)

File: mail_sendgrid_mass_mailing/__manifest__.py
Comment (simahawk) [inline mail_sendgrid_mass_mailing/__manifest__.py]: https://github.com/OCA/social

File: mail_sendgrid_mass_mailing/models/mass_mailing.py
Comment (simahawk) [inline mail_sendgrid_mass_mailing/models/mass_mailing.py]: better to split this to `_send_mail_get_composer_values` or alike

File: mail_sendgrid_mass_mailing/tests/test_mass_mailing.py
Comment (simahawk) [inline mail_sendgrid_mass_mailing/tests/test_mass_mailing.py]: same: try using `setUpClass`

**Review by simahawk — CHANGES_REQUESTED** [review]
wow, a long one :sweat_smile:

File: mail_sendgrid/models/email_template.py
Comment (ecino) [reply mail_sendgrid/models/email_template.py]: +1 for defaultdict, I didn't new this library!

File: mail_sendgrid/models/mail_mail.py
Comment (ecino) [reply mail_sendgrid/models/mail_mail.py]: I thought write method should not be used in compute methods?

Comment (hbrunn) [reply mail_sendgrid/models/mail_mail.py]: it absolutely shouldn't and wouldn't work here anyways. What the colleague probably means is [.update()](https://github.com/OCA/OCB/blob/10.0/odoo/models.py#L5010)

File: mail_sendgrid/models/mail_mail.py
Comment (ecino) [reply mail_sendgrid/models/mail_mail.py]: search instead of mapped? for better performance?

File: mail_sendgrid/wizards/email_template_preview.py
Comment (ecino) [reply mail_sendgrid/wizards/email_template_preview.py:14]: No, because the parent method in odoo/addons/mail has this name

File: mail_sendgrid/tests/test_mail_sendgrid.py
Comment (ecino) [reply mail_sendgrid/tests/test_mail_sendgrid.py]: Not working with HttpCase? I get errors that cls.env is not defined

File: mail_sendgrid_mass_mailing/tests/test_mass_mailing.py
Comment (ecino) [reply mail_sendgrid_mass_mailing/tests/test_mass_mailing.py]: I tried with it but I get many errors in tearDown (DB no such savepoint)

**Comment by ecino** [comment]
@simahawk Thank you very much for your helpful review! I corrected most of what you found.
@pedrobaeza I updated the tests with more coverage. Unfortunately the coverage is not yet enough... very hard to match 98.8%!

File: mail_sendgrid/wizards/email_template_preview.py
Comment (simahawk) [reply mail_sendgrid/wizards/email_template_preview.py:14]: sorry, my bad, I missed the `super` call ;P

File: mail_sendgrid_mass_mailing/tests/test_mass_mailing.py
Comment (simahawk) [reply mail_sendgrid_mass_mailing/tests/test_mass_mailing.py]: you have to use `common.SavepointCase` as base class instead.

Comment (ecino) [reply mail_sendgrid_mass_mailing/tests/test_mass_mailing.py]: Yes I tried that, but then I don't know what the errors are about in the tearDown method.

Comment (simahawk) [reply mail_sendgrid_mass_mailing/tests/test_mass_mailing.py]: which errors? do you have a traceback?

Comment (ecino) [reply mail_sendgrid_mass_mailing/tests/test_mass_mailing.py]: Ok, that was my bad, I forgot to add the `@classmethod` decorators and change tearDown to tearDownClass as well.

I still couldn't change the test of Sendgrid because of the `HttpCase` class.

**Comment by rafaelbn** [comment]
@simahawk please could you unblock this PR? :smile:

**Comment by rafaelbn** [comment]
If there is not any inconvenient I will merge this today evening.

**Review by simahawk — APPROVED** [review]
@ecino @rafaelbn sorry, went off my radar :stuck_out_tongue_winking_eye:

**Comment by ecino** [comment]
@rafaelbn Can we merge?

**Comment by pedrobaeza** [comment]
Can you squash a bit the commit history before merging? I see several commits with same text and other not very relevant.

**Comment by ecino** [comment]
@pedrobaeza Done

**Pull request merged**
